src/alchemlyb/tests/test_preprocessing.py

```diff
@@ -213,6 +236,97 @@ def test_raise_ValueError_for_mismatched_data(self, series):
         with pytest.raises(ValueError):
             self.slicer(data, series=series)
 
+    @pytest.mark.parametrize(('data', 'lower', 'upper'),
+                             [
+                                 (gmx_benzene_dHdl(), 1000, 34000),
+                                 (gmx_benzene_u_nk(), 1000, 34000),
+                             ])
+    def test_lower_and_upper_bound_slicer(self, data, lower, upper):
+        """
+        Test that the lower and upper time is respected when using statistical_inefficiency
+        without a series. In this case, statistical_inefficiency should behave like slicing
+        """
+        original_length = len(data)
+        # Check that the input data is appropriate for the test
+        assert any(data.reset_index()['time'] < lower)
+        assert any(data.reset_index()['time'] > upper)
+
+        # Slice data, and check that we don't observe times outside
+        # the prescribed range
+        sliced = self.slicer(data,
+                             series=None,
+                             lower=lower,
+                             upper=upper,
+                             step=5)
+        assert all(sliced.reset_index()['time'] >= lower)
+        assert all(sliced.reset_index()['time'] <= upper)
+
+        # Make sure we didn't change input data
+        assert len(data) == original_length
+
+    @pytest.mark.parametrize(('data', 'lower', 'upper'),
+                             [
+                                 (gmx_benzene_dHdl(), 1000, 34000),
+                                 (gmx_benzene_u_nk(), 1000, 34000),
+                             ])
+    def test_lower_and_upper_bound_inefficiency(self, data, lower, upper):
+        """
+        Test that the lower and upper time is respected when using statistical_inefficiency
+        with a series. In this case, statistical_inefficiency should slice the series, then
+        subsample the data frame.
+        """
+        original_length = len(data)
+        # Check that the input data is appropriate for the test
```

Comment:
I think you don't need this for every tests.
Might create a test called `test_sanity` and check for this.

Reply:
Here's my rationale for implementing this on every test:

There is at least three ways to render this test useless:
* Someone changes the input data to this test from `gmx_benzene_*()` to something else
* Someone changes the lower / upper bound used in this test
* Someone changes the underlying data that `gmx_benzene_*()` returns

The solution you suggest addresses my third point. But then you'd need a check to make sure that the two tests are using the same input data...

Additionally, this check is also an information for developers looking at this test - we are clearly stating what we are expecting coming in, so the rest of the test is easier to understand.

Reply:
makes sense; I'd repackage as a function (see my other comment)

Reply:
@ptmerz 

> Someone changes the underlying data that gmx_benzene_*() returns

I do agree with this one though it might be very less likely. a separate sanity test should be good enough.

> Someone changes the lower / upper bound used in this test

Do you mind elaborating on this? I think in general there is no need to change the test unless it failed but that would be a different case.

> Someone changes the input data to this test from gmx_benzene_*() to something else
I don't think you need to guard this as I cannot think of a user case where this would happen?

Reply:
> @ptmerz
> 
> > Someone changes the underlying data that gmx_benzene_*() returns
> 
> I do agree with this one though it might be very less likely. a separate sanity test should be good enough.

Sure, that's the case that would be covered by a separate test.

> > Someone changes the lower / upper bound used in this test
> 
> Do you mind elaborating on this? I think in general there is no need to change the test unless it failed but that would be a different case.

I am not referring to needing to changing the tests. My guiding principle when writing tests is to make them as secure as possible. Bugs are usually not introduced on purpose, but because someone misunderstood what a functionality (or a test) does, or someone was not aware of side effects that a change has in other parts of the code. This `assert` guards from someone mistakenly changing the input data of the test to something that stops this test from being useful.

> > Someone changes the input data to this test from gmx_benzene_*() to something else
> 
> I don't think you need to guard this as I cannot think of a user case where this would happen?

Again, that's not a user case. It could very well happen that in a while a developer comes up with a great new input data set to be used in tests. With the test implemented as I am suggesting, that person can confidently change the input data without having to understand the details of the test. Imagine we didn't have these additional tests. They might want to add shorter input data to speed up the run time of tests. They could add that and the test would still pass - after all, `assert all(sliced.reset_index()['time'] <= upper)` is going to always pass if the input data was shorter than the upper bound from the start.

Reply:
Maybe some of the confusion comes from the lack of syntactic difference between checks to assert that certain assumptions that I (as test writer) made are correct, and the actual test of functionality. In natural language, this would go something like that:

**Given** a dataset and upper and lower bounds,  
**assuming that** the input data set has data that is lower than the lower bound and higher than the upper bound,  
**test that** `slicing` does successfully remove frames before the lower bound and after the higher bound, and does not change the input file.

The important part is that the test is useless if the assumptions don't hold. I am not making any guesses about how probable it is that someone changes the test files in a way that these assumptions fail.

Another way to look at it is from the point of view of the function. The test function is principally a function that says "I will assert that `slicing` works correctly". There's two ways to interpret this:
1. I will assert that `slicing` works if and only if you give me reasonable input data. Otherwise, my behavior is undefined (because I might still tell you that everything works although I haven't actually been able to check anything). 
2. I will assert that `slicing` works or loudly complain if I'm unable to perform that task with the input data you gave me.

I prefer the second approach, so that's what I went for.